digits/model/tasks/caffe_train.py

```diff
@@ -860,7 +860,10 @@ def _pycaffe_args(self, gpu_id):
             solver_type = solver_type_mapping[self.solver_type]
         except KeyError:
             raise ValueError("Unknown solver type {}.".format(self.solver_type))
-        gpu_script = "caffe.set_device({id});".format(id=gpu_id) if gpu_id else ""
+        if gpu_id:
+            gpu_script = "caffe.set_device({id});caffe.set_mode_gpu();".format(id=gpu_id)
+        else:
+            gpu_script = "caffe.set_mode_gpu();"
```

Comment:
My bad.  I didn't explain that portion clearly.  When gpu_id == None, it means either the resources['gpu'] is empty or 'gpu' is not in resources.  Those two conditions do not set mode to CPU in regular path of task_arguments.  I tried to match that behavior in _pycaffe_args.